src/utils/digest.ts

```diff
@@ -0,0 +1,88 @@
+// that can be found in the LICENSE file exposed on Github (readium) in the project repository.
+// ==LICENSE-END==
+
+import { createHash } from "crypto";
+
+function decimalToHex(d: number, padding: number) {
+    let hex = Number(d).toString(16);
+    padding = typeof (padding) === "undefined" || padding === null ? padding = 2 : padding;
+
+    while (hex.length < padding) {
+        hex = "0" + hex;
+    }
+
+    return hex;
+}
+
+export const parseHeaderWWWAuthenticateForDigest = (v: string): { [s: string]: any } => {
+    const a = v.trim().split(",");
+    const b = a.map((q) => q.trim().split("="));
+    const c = b.map((w) => w.length != 2 ? undefined : w.map((e) => e.trim().replace(/(^"|"$)/g, "")));
+    const d = c.filter((r) => !!r);
+    const e = d.reduce((pv, cv) => ({ ...pv, [cv[0]]: cv[1] }), {});
+    return e;
+};
+
+export interface IDigestDataParsed {
+    username: string,
+    realm: string,
+    nonce: string,
+    algorithm: "MD5" | "MD5-sess" | undefined,
+    qop: "auth" | "auth-int" | undefined,
+    cnonce: string
+    uri: string,
+    nonceCount: string,
+}
+
+export const parseDigestString = (wwwAuthenticate: string): IDigestDataParsed => {
+
+    const data = parseHeaderWWWAuthenticateForDigest(wwwAuthenticate);
+    return {
+        username: data["username"] || "",
+        realm: data["realm"] || "",
+        nonce: data["nonce"] || "",
+        algorithm: data["algorithm"] === "MD5" ? "MD5" : data["algorithm"] === "MD5-sess" ? "MD5-sess" : undefined,
+        qop: data["qop"] === "auth" ? "auth" : data["qop"] === "auth-int" ? "auth-int" : undefined,
+        cnonce: data["cnonce"],
+        uri: data["uri"],
+        nonceCount: data["nonceCount"],
+    };
+};
+
+export const digestAuthentication = ({
+    username,
+    password,
+    nonce,
+    qop,
+    algorithm,
+    realm,
+    cnonce,
+    uri,
+    method,
+    nonceCount,
+}: {
+    username: string,
+    password: string,
+    nonce: string,
+    qop: string,
+    algorithm: string,
+    realm: string,
+    cnonce: string,
+    uri: string,
+    method: string,
+    nonceCount: string,
+}) => {
+
+    nonceCount = nonceCount ? decimalToHex(parseInt(nonceCount, 16) + 1, 8) : decimalToHex(1, 8);
+    const ha1MD5 = createHash("md5").update(`${username}:${realm}:${password}`).digest("hex");
+    const ha1 = algorithm === "MD5-sess" ? createHash("md5").update(`${ha1MD5}:${nonce}:${cnonce}`).digest("hex") : ha1MD5;
+    const ha2 = createHash("md5").update(qop === "auth-int" ? "" : `${method}:${uri}`).digest("hex"); // qop === "auth-int" not supported what is entityBody?
+    const response = createHash("md5").update((qop === "auth" || qop === "auth-int") ? `${ha1}:${nonce}:${nonceCount}:${cnonce}:${qop}:${ha2}` : `${ha1}:${nonce}:${ha2}`).digest("hex");
+    const accessToken = `username="${username}", realm="${realm}", nonce="${nonce}", qop=${qop}, algorithm=${algorithm}, response="${response}", uri="${uri}", nc=${nonceCount}, cnonce="${cnonce}"`;
```

Comment:
The key/value pair `username="${username}"` is incorrect if `username` contains characters forbidden in the "quoted string" syntax, or more generally Unicode characters. Instead, `username*=` must be used.
Suggestion: maybe consider passing `userhash=true` and normalise `username` via a hashing function that generates a valid "quoted string"? (not sure about Calibre server support though...)

See:
https://httpwg.org/specs/rfc7616.html#rfc.section.3.4
Unicode example:
> username*=UTF-8''J%C3%A4s%C3%B8n%20Doe

Reply:
there are no implementation of `username*=` or `userhash=` in calibre I think

https://github.com/kovidgoyal/calibre/blob/eb78a761a99ac20a6364f85e12059fec6517d890/src/calibre/srv/tests/auth.py

Reply:
they used https://www.rfc-editor.org/rfc/rfc2617

Reply:
Resolving this issue as a no-fix, for now. As this is a known implementation caveat, I am adding a code comment to record the information, in case the bug surfaces in the real world in the future.

Reply:
See 21f490f73bccff38a786b172fb584cb58d84a3c3